pkg/antctl/commandlist.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2019 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package antctl
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"math"
+	"net/http"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/spf13/cobra"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apiserver/pkg/server/mux"
+	"k8s.io/client-go/util/homedir"
+	"k8s.io/klog"
+
+	"github.com/vmware-tanzu/antrea/pkg/monitor"
+)
+
+// commandList organizes definitions.
+// It is the protocol for a pair of antctl client and server.
+type commandList struct {
+	definitions  []commandDefinition
+	groupVersion *schema.GroupVersion
+	codec        serializer.CodecFactory
+}
+
+// apiPrefix returns the API prefix of the antctl service.
+func (cl *commandList) apiPrefix() string {
+	return path.Join("/apis", cl.groupVersion.Group, cl.groupVersion.Version)
+}
+
+// ApplyToMux adds the handler function of each commandDefinition in the
+// commandList to the mux with path /<api prefix>/<cmd>, it also adds a dummy discovery
+// handler with path /<api prefix>. Since the antctl service contains only non-resource
+// data, we need setup the discovery handler here to satisfy the apiserver.
+func (cl *commandList) ApplyToMux(mux *mux.PathRecorderMux, aq monitor.AgentQuerier, cq monitor.ControllerQuerier) {
+	for _, def := range cl.definitions {
+		handler := def.HandlerFactory.Handler(aq, cq)
+		reqPath := def.requestPath(cl.apiPrefix())
+		klog.Infof("Adding cli handler %s", reqPath)
+		mux.HandleFunc(reqPath, handler)
+	}
+	// Set up the discovery handler.
```

Comment:
Since it seems to be a key function for antctl, we can enhance the comments for this one. Maybe change the comment to "Set up the dummy discovery handler" in this line to be consistent. "satisfy the apiserver" sounds vague and it will be better to add what the API Server requirements are.

Reply:
Updated, PTAL.